# 01. Importing libraries

In [1]:
import pandas as pd
import numpy as np
import os

# 02. Importing data

In [2]:
path = r'/Users/marcela/Achievement 4'

In [3]:
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [4]:
# Create a subset 

In [5]:
df = ords_prods_merge[:1000000]

# 03. Creating an If statement with a User-Defined function

In [6]:
# Define a function
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [7]:
# Apply the function
df['price_range'] = df.apply(price_label, axis=1)

/var/folders/gz/793390fs5lz_gt8d667yd0qw0000gn/T/ipykernel_54692/902492192.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [34]:
# Create a copy of df if it's a slice from another DataFrame
df = df.copy()

# Define the function
def price_label(row):
    if row['prices'] <= 5:
        return 'Low-range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid-range product'
    elif row['prices'] > 15:
        return 'High range'
    else: 
        return 'Not enough data'

# Apply the function
df['price_range'] = df.apply(price_label, axis=1)

In [35]:
df.shape

(1000000, 16)

In [36]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [37]:
df['prices'].max()

14.8

# 04. Creating an If statement with the loc() Function

In [38]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [39]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [40]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [41]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [42]:
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [43]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <= 15) & (ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [44]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [45]:
ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21890146
Low-range product     10126384
High-range product      417682
Name: count, dtype: int64

In [46]:
# Checking the frequency of the 'orders_day_of_week' column
ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6209632
1    5665830
6    4500246
2    4217766
5    4209449
3    3844096
4    3787193
Name: count, dtype: int64

# 05. Creating an If statement with For-Loops

In [47]:
result = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [48]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [49]:
ords_prods_merge['busiest_day'] = result

In [50]:
ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22437387
Busiest day        6209632
Least busy         3787193
Name: count, dtype: int64

# Task Step 2. Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method

In [51]:
results = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0 or value == 1:
    results.append("Busiest day")
  elif value == 4 or value == 3:
    results.append("Least busy")
  else:
    results.append("Regularly busy")

In [52]:
ords_prods_merge['busiest_day'] = results

In [53]:
ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    12927461
Busiest day       11875462
Least busy         7631289
Name: count, dtype: int64

# Task Step 3. Check the values of this new column for accuracy. Note any observations in markdown format

In [54]:
ords_prods_merge.shape

(32434212, 17)

## The total is 32,434,212.

# Task Step 4. When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [55]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day busiest_period_of_day  
0  Regularly busy        Average orders  
1      Least busy        Average orders  
2      Least busy        Average orders  
3      Least busy        Average orders  
4      Least busy        Average orders

In [56]:
ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2764390
11    2738585
14    2691563
15    2664522
13    2663272
12    2620800
16    2537469
9     2456661
17    2089452
8     1719952
18    1637922
19    1259382
20     977017
7      891928
21     796362
22     634737
23     402612
6      290770
0      218942
1      115786
5       88057
2       69431
4       53283
3       51317
Name: count, dtype: int64

In [57]:
hour = []

for value in ords_prods_merge["order_hour_of_day"]:
  if value == 10 or value == 11 or value == 14:
    hour.append("Most orders")
  elif value == 2 or value == 4 or value == 3:
    hour.append("Fewest orders")
  else:
    hour.append("Average orders")

In [58]:
ords_prods_merge['busiest_period_of_day'] = hour

# Task Step 5. Print the frequency for this new column

In [59]:
ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Average orders    24065643
Most orders        8194538
Fewest orders       174031
Name: count, dtype: int64

# Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder

In [60]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_busiestdays.pkl'))